core/src/main/java/org/infinispan/interceptors/distribution/L1LastChanceInterceptor.java

```diff
@@ -82,44 +97,112 @@ public Object visitComputeIfAbsentCommand(InvocationContext ctx, ComputeIfAbsent
+   }
+
+   @Override
+   public Object visitReadWriteManyEntriesCommand(InvocationContext ctx, ReadWriteManyEntriesCommand command) throws Throwable {
+      return invokeNextThenApply(ctx, command, handleWriteManyCommand);
+   }
+
    public Object visitDataWriteCommand(InvocationContext ctx, DataWriteCommand command, boolean assumeOriginKeptEntryInL1) throws Throwable {
-      return invokeNextThenAccept(ctx, command, (rCtx, rCommand, rv) -> {
-         Object key;
-         DataWriteCommand writeCommand = (DataWriteCommand) rCommand;
-         Object key1 = (key = writeCommand.getKey());
-         if (shouldUpdateOnWriteCommand(writeCommand) && writeCommand.isSuccessful() &&
-               cdl.getCacheTopology().isWriteOwner(key1)) {
-            if (trace) {
-               log.trace("Sending additional invalidation for requestors if necessary.");
-            }
-            // Send out a last attempt L1 invalidation in case if someone cached the L1
-            // value after they already received an invalidation
-            blockOnL1FutureIfNeeded(l1Manager
-                  .flushCache(Collections.singleton(key), rCtx.getOrigin(), assumeOriginKeptEntryInL1));
+      return invokeNextThenApply(ctx, command, assumeOriginKeptEntryInL1 ? handleDataWriteCommandEntryInL1 : handleDataWriteCommandEntryNotInL1);
+   }
+
+   private Object handleDataWriteCommand(InvocationContext rCtx, VisitableCommand rCommand, Object rv, boolean assumeOriginKeptEntryInL1) {
+      Object key;
+      DataWriteCommand writeCommand = (DataWriteCommand) rCommand;
+      Object key1 = (key = writeCommand.getKey());
+      if (shouldUpdateOnWriteCommand(writeCommand) && writeCommand.isSuccessful() &&
+            cdl.getCacheTopology().isWriteOwner(key1)) {
+         if (trace) {
+            log.trace("Sending additional invalidation for requestors if necessary.");
          }
-      });
+         // Send out a last attempt L1 invalidation in case if someone cached the L1
+         // value after they already received an invalidation
+         CompletableFuture<?> f = l1Manager.flushCache(Collections.singleton(key), rCtx.getOrigin(), assumeOriginKeptEntryInL1);
+         return asyncReturnValue(f, rv);
+      }
+      return rv;
+   }
+
+   private Object handleDataWriteCommandEntryInL1(InvocationContext rCtx, VisitableCommand rCommand, Object rv) {
+      return handleDataWriteCommand(rCtx, rCommand, rv, true);
+   }
+
+   private Object handleDataWriteCommandEntryNotInL1(InvocationContext rCtx, VisitableCommand rCommand, Object rv) {
+      return handleDataWriteCommand(rCtx, rCommand, rv, false);
+   }
+
+   private Object asyncReturnValue(CompletableFuture<?> f, Object rv) {
+      if (f == null || f.isDone()) {
+         return rv;
+      }
+      return asyncValue(f.handle((nil, throwable) -> {
+         // Ignore SuspectExceptions - if the node has gone away then there is nothing to invalidate anyway.
```

Comment:
The comment still mentions SuspectException`.

TBH I didn't read the code close enough, I just assumed exceptions other than `SuspectException` fail the write because one node now sees a stale value, and that you'd be able to remove `asyncReturnValue` altogether. WDYT?